File: 450. Delete Node in a BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        # edge case
        if not root:
            return root
        
        # attach a dummy root
        dummy = TreeNode(sys.maxsize)
        dummy.left = root
        prev = dummy
        
        # find the node
        node = root

        while node and node.val != key:
            prev = node
            if node.val < key:
                node = node.right
            elif node.val > key:
                node = node.left

        # check if found
        if not node:
            return root
        
        # adjust subtrees
        left = node.left
        right = node.right
        subtree = None
        if right:
            subtree = right
            if left:
                ptr = right
                while ptr.left:
                    ptr = ptr.left
                ptr.left = left
        else:
            subtree = left
            
        # remove the node
        if prev.left == node:
            prev.left = subtree
        else:
            prev.right = subtree
            
        return dummy.left
                
            
        

File: 450. Delete Node in a BST/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        
        // add a dummy node
        TreeNode node = new TreeNode(Integer.MAX_VALUE);
        node.left = root;
        root = node;
        node = root.left;
        TreeNode prev = root;
        
        // search for the key
        while (node != null) {
            if (node.val == key) {
                break;
            } else if (node.val > key) {
                prev = node;
                node = node.left;
            } else {
                prev = node;
                node = node.right;
            }
        }
        
        // key found
        if (node != null) {
            TreeNode tmp = null;
            
            if (node.left == null && node.right == null) {
                ;
            } else if (node.left == null && node.right != null) {
                tmp = node.right;
            } else if (node.left != null && node.right == null) {
                tmp = node.left;
            } else {
                tmp = node.right;
                while (tmp.left != null) {
                    tmp = tmp.left;
                }
                
                tmp.left = node.left;
                tmp = node.right;
            }
        
            if (prev.left == node) {
                prev.left = tmp;
            } else {
                prev.right = tmp;
            }
        }
        
        return root.left;
    }
}
